core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -7330,4 +7312,58 @@ private boolean isInternalRole(String domain) {
 
         return domain.equals("Internal") || domain.equals("Application");
     }
+
+    private String removeDomainIfNotApplicationRole(String roleName) {
+
+        String formattedRoleName;
+        if (roleName.contains(UserCoreConstants.DOMAIN_SEPARATOR)
+                && roleName.toLowerCase().startsWith(APPLICATION_DOMAIN.toLowerCase())) {
+            formattedRoleName = roleName;
+        } else {
+            formattedRoleName = UserCoreUtil.removeDomainFromName(roleName);
+        }
+        return formattedRoleName;
+    }
+
+    private void createHybridRole(String roleName, String[] userList, org.wso2.carbon.user.api.Permission[] permissions)
+            throws UserStoreException {
+
+        // It is possible that the adding role could already exists at the table. But if concurrent requests were made,
+        // it is possible that the adding role does not exists at this moment, but it still could exists at the
+        // moment when DB query is called.
+        try {
+            hybridRoleManager.addHybridRole(roleName, userList);
+        } catch (UserStoreException e) {
+            // Could be already existing error or unique constraint violation, either of them could point to already
+            // existing role issue.
+            if (hybridRoleManager.isExistingRole(roleName)) {
```

Comment:
Shouldn't we check an error code or something before we do this lookup?
otherwise, we will end up doing this lookup for all exception scenarios, isn't it?

Reply:
Addressed in 811cae82352b6311070094f9dafcb67c025b42af.